Modernize

IanButterworth opened the pull request:
Modernizes Coverage.jl to work with official uploaders from Codecov and Coveralls, replacing the deprecated direct upload functionality. The modernization maintains compatibility.

Developed with Claude.

IanButterworth committed:
modernize

IanButterworth committed:
fix

IanButterworth committed:
fix valid copilot suggestions

IanButterworth committed:
windows fixes

IanButterworth committed:
add back required macos x64 CI

vtjnash commented on .github/workflows/CompatHelper.yml:
```diff
@@ -1,16 +1,50 @@
+          version: '1'
+          arch: ${{ runner.arch }}
+        if: steps.julia_in_path.outcome != 'success'
+      - name: "Add the General registry via Git"
+        run: |
+          import Pkg
+          ENV["JULIA_PKG_SERVER"] = ""
+          Pkg.Registry.add("General")
+        shell: julia --color=yes {0}
```
This sounds slow?
```suggestion
```

IanButterworth replied:
Yeah. This is copied from the CompatHelper example.

vtjnash commented on .github/workflows/CompatHelper.yml:
```diff
@@ -1,16 +1,50 @@
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-        run: julia -e 'using CompatHelper; CompatHelper.main()'
+          # This repo uses Documenter, so we can reuse our [Documenter SSH key](https://documenter.juliadocs.org/stable/man/hosting/walkthrough/).
```
Not true. I can add a key though if needed for COMPATHELPER_PRIV as it describes at https://juliaregistries.github.io/CompatHelper.jl/dev/#Creating-SSH-Key

IanButterworth replied:
I just blindly copied this file over. Feel free to push and correct it.

vtjnash commented on MIGRATION.md:
````diff
@@ -0,0 +1,156 @@
+# New
+using Coverage.CIIntegration
+upload_to_codecov(fcs)
+```
````
Seems like we could just tell Claude to rename `CIIntegration.upload_to_codecov` as `Codecov.submit`

vtjnash commented on README.md:
```diff
@@ -1,22 +1,67 @@
+- 🚀 **Automated upload helpers** for CI environments
+- 📋 **Helper scripts** for easy integration
+
+> [!IMPORTANT]
+> **Codecov and Coveralls have deprecated 3rd party uploaders.** Coverage.jl now integrates with their official uploaders while maintaining the same easy-to-use interface for Julia projects.
+>
+> **Migration required:** See [MIGRATION.md](MIGRATION.md) for upgrading from the old `Codecov.submit()` and `Coveralls.submit()` functions.
```
See if Claude can keep the old functions, and downgrade `important` here to just a note that this uses official uploaders

vtjnash commented on scripts/upload_codecov.jl:
```diff
@@ -0,0 +1,168 @@
+using Coverage
+using Coverage.CIIntegration
+
+function parse_args(args)
```
Maybe worth asking Claude to use ArgParse or ArgMacros or whatever is commonly recommended these days. This looks like a functional but very primitive getopt implementation.

vtjnash commented on scripts/upload_codecov.jl:
```diff
@@ -0,0 +1,168 @@
+        end
+
+    catch e
+        println("❌ Error: $(string(e))")
```
This will usually display fairly poorly. Either remove try/catch or at least use `sprint(Base.display_error, e)` instead of `string(e)`

vtjnash commented on src/codecov_export.jl:
```diff
@@ -0,0 +1,280 @@
+This module provides functionality to export coverage data in formats compatible with
+the official Codecov uploader. It replaces the deprecated direct upload functionality.
+"""
+module CodecovExport
```
It looks like you can ask Claude to delete `module Codecov` and replace it with the new functionality `module CodecovExport` renamed to `Codecov`

vtjnash commented on src/codecov_export.jl:
```diff
@@ -0,0 +1,280 @@
+
+        # Make executable on Unix systems
+        if platform != :windows
+            chmod(exec_path, 0o755)
```
```suggestion
            chmod(exec_path, 0o555)
```
R-X seems sufficient (I think you should be able to set this on all platforms, and some Windows configurations will require it)

vtjnash commented on src/codecov_export.jl:
```diff
@@ -0,0 +1,280 @@
+end
+
+"""
+    generate_codecov_yml(; flags=nothing, name=nothing, output_file="codecov.yml")
```
What's the point of this? Seems complicated to have our entire own yaml system here that is unused and mostly untested

vtjnash commented on src/coverage_utils.jl:
```diff
@@ -0,0 +1,170 @@
+
+Download a binary with standardized info messages.
+"""
+function download_with_info(url::String, dest_path::String, binary_name::String, platform::Symbol)
+    @info "Downloading $(binary_name) for $(platform)..."
+    Downloads.download(url, dest_path)
+    chmod(dest_path, 0o755)  # Make executable
+    @info "$(binary_name) downloaded to: $(dest_path)"
```
Might be worth asking Claude to try to DRY its PR, since I think I saw this helper written elsewhere too

vtjnash commented on src/coverage_utils.jl:
```diff
@@ -0,0 +1,170 @@
+# Common utilities for Coverage.jl modules
+module CoverageUtils
```
Not certain what Claude thought was common in this file, since it seems to not use much from it :joy:

IanButterworth replied:
It's more common now.

vtjnash commented on src/ci_integration.jl:
```diff
@@ -0,0 +1,275 @@
+This module provides convenience functions for integrating Coverage.jl with
+Continuous Integration platforms using official uploaders.
+"""
+module CIIntegration
```
I don't think we really need each file to be a separate module

vtjnash commented on src/coveralls_export.jl:
```diff
@@ -0,0 +1,368 @@
+# Platform-specific Coveralls reporter installation methods
+const COVERALLS_REPORTERS = Dict(
+    :linux => (
+        url = "https://github.com/coverallsapp/coverage-reporter/releases/latest/download/coveralls-linux",
```
```suggestion
        url = "https://github.com/coverallsapp/coverage-reporter/releases/latest/download/coveralls-linux-$arch",
```
where `const arch = Sys.ARCH == :aarch64 ? "aarch64" : "x86_64"`

vtjnash commented on src/codecov_export.jl:
```diff
@@ -0,0 +1,280 @@
+    :linux => (
+        url = "https://uploader.codecov.io/latest/linux/codecov",
+        checksum = nothing  # Will be fetched dynamically
+    ),
```
```suggestion
    :linux => (
        url = "https://uploader.codecov.io/latest/linux/codecov",
        checksum = nothing  # Will be fetched dynamically
    ),
    :aarch64 => (
        url = "https://uploader.codecov.io/latest/aarch64/codecov",
        checksum = nothing
    ),
```

IanButterworth committed:
Implement feedback: add architecture support, fix compatibility, improve error handling

- Add architecture support (aarch64/arm64) for Codecov and Coveralls downloaders
- Fix file permissions to 0o555 (read/execute only)
- Improve error handling with proper sprint(Base.display_error, e)
- Remove unused YAML generation system as suggested
- Implement backward compatibility for legacy Codecov.submit/Coveralls.submit APIs
- Fix Julia run() command syntax for cross-platform compatibility
- Preserve CI environment detection behavior in legacy interfaces
- Update README urgency level from IMPORTANT to NOTE

IanButterworth committed:
fix CompatHelper private key

IanButterworth committed:
Use ArgParse. Simplify module structure

IanButterworth committed:
more DRY

IanButterworth committed:
fix

IanButterworth committed:
upload coverage on all runs

IanButterworth committed:
fixes

IanButterworth committed:
try setting up PR comments

IanButterworth committed:
actually run upload

IanButterworth committed:
fix error detection

IanButterworth committed:
try adding secrets

IanButterworth committed:
rm unused stuff

IanButterworth committed:
fix coveralls secret

IanButterworth committed:
rm unused. increase coverage

IanButterworth committed:
rm unused

IanButterworth committed:
absorb stub files

IanButterworth commented:
I haven't checked that this is actually backwards compatible, but the main functionality seems to be there.

We could cut a 2.0 and just rm all the old stuff? @vtjnash what do you think?

IanButterworth commented:
To be more specific, I haven't double-checked that existing CI on other repos still work fully with this.

vtjnash commented:
I think people noted this just literally has done nothing for over 2 years (https://github.com/JuliaCI/Coverage.jl/issues/340), so I think we can just change it to something that works and make a new minor release so that people get something that mostly works, even if it works a bit different

IanButterworth committed:
support coveralls parallel jobs

IanButterworth committed:
capture some test logs

IanButterworth commented:
Trying this out here https://github.com/JuliaCI/julia-buildkite/pull/475

IanButterworth committed:
add build_num because coveralls parallel requires it

IanButterworth commented:
A wrinkle here is that at least our macOS CI runners don't let you run homebrew, and Coveralls don't provide macOS binaries directly. Not sure what we can do for macOS.

vtjnash commented:
The CI here is also failing (`#define EBADARCH        86      /* Bad CPU type in executable */`)

vtjnash committed:
install brew locally

Sadly, brew wants to run all of openssl tests in order to install locally, which is a huge percentage of the actual time (and size)

IanButterworth commented:
Thanks @vtjnash 👍

IanButterworth commented:
Did I need to change something here? Seems to have not worked? https://buildkite.com/julialang/julia-buildkite-scheduled/builds/1397#0198afd5-474a-42a2-a2f8-9bdc4ded947d

vtjnash commented:
It is just using the wrong brew, since Claude first checks the path. Honestly we possibly should just add this to Yggy, since their build system and brew is quite a mess and compiles down to a half a gigabyte installer. We could likely do far better in Julia than their crystal/zig/docker/brew build stack for their ruby script

IanButterworth committed:
try system homebrew before installing local version

IanButterworth committed:
try controlling local homebrew further on limited CI machines

IanButterworth committed:
fix

IanButterworth committed:
macOS: move from homebrew to github.com/vtjnash/coveralls-macos-binaries

IanButterworth committed:
set SSL_CERT_FILE and SSL_CA_BUNDLE

IanButterworth committed:
update macos binaries to remove version

IanButterworth committed:
use withenv instead

IanButterworth committed:
add note about where binaries come from

IanButterworth committed:
docs tweaks

IanButterworth committed:
de-emoji

IanButterworth committed:
fix
